pkg/metrics/v2/sender.go

```diff
@@ -0,0 +1,263 @@
+package metrics
+
+import (
+	"fmt"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/pkg/metrics/wal"
+	"github.com/grafana/agent/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	prom_config "github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/scrape"
+	prom_storage "github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/prometheus/statsd_exporter/pkg/level"
+)
+
+type senderMetrics struct {
+	util.MetricsContainer
+
+	numberSenders       prometheus.Gauge
+	totalSenderLookups  prometheus.Counter
+	failedSenderLookups prometheus.Counter
+}
+
+func newSenderMetrics() *senderMetrics {
+	var m senderMetrics
+
+	m.numberSenders = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "agent_metrics_senders_count",
+		Help: "Current number of running senders",
+	})
+	m.totalSenderLookups = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "agent_metrics_sender_lookups_total",
+		Help: "Total number of sender lookups (successful and failed)",
+	})
+	m.failedSenderLookups = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "agent_metrics_sender_failed_lookups_total",
+		Help: "Total number of failed sender lookups",
+	})
+
+	m.Add(
+		m.numberSenders,
+		m.totalSenderLookups,
+		m.failedSenderLookups,
+	)
+	return &m
+}
+
+// storageSet abstracts over a series of WALs by name. Returns an error if the
+// provided name isn't known.
+type storageSet interface {
+	Appendable(name string) (storage, error)
+}
+
+type storage interface {
+	prom_storage.Appendable
+
+	// bindScraper binds the scraper to the storage. This is used for sending
+	// metadata.
+	bindScraper(sm *scrape.Manager)
```

Comment:
This was actually a pretty big issue I ran into. If we don't attach the scrape manager to remote write, then we don't get metadata for any of our scraped metrics (help/text info). 

Unfortunately, this also means that v2 integrations aren't generating any metadata because they're being scraped by an external scraper. I'm not sure how we should handle that yet, but I'll open up an issue for tracking.

Reply:
#1262